File: leetcode/src/main/ruby/longest_palindrome.rb
# @param {String} s
# @return {String}
def longest_palindrome(s)
    return s if s.length <= 1
    t = pre_process(s)

    idx, right, m_idx = 0, 0, 0
    p = Array.new(t.length, 0)

    for i in 1...t.length - 1
        p[i] = i < right ? [p[2 * idx - i] , right - i].min : 0
        p[i] += 1 while (t[i - p[i] - 1] == t[i + p[i] + 1])
        if (i + p[i] > right)
            idx = i
            right = i + p[i]
        end
        m_idx = i if p[i] >= p[m_idx]
    end

    start = (m_idx - p[m_idx] - 1) / 2
    s[start...start + p[m_idx]]
end

def pre_process(s)
    t = Array.new(s.length * 2 + 3)
    i = -1
    t.each_with_index do |_, idx|
        t[idx] =
            if idx == 0
                '^'
            elsif idx == t.length - 1
                '$'
            elsif idx % 2 == 1
                '#'
            else
                i += 1
                s[i]
            end
    end
end

File: leetcode/src/main/golang/longest_palindrome.go
func longestPalindrome(s string) string {
	if len(s) <= 1 {
		return s
	}

	t := preProcess(s)
	p := make([]int, len(t))
	idx, right, mIdx := 0, 0, 0

	for i := 1; i < len(t) - 1; i++ {
		p[i] = 0
		if i < right {
			p[i] = min(p[2 * idx - i], right - i)
		}
		for ; t[i + p[i] + 1] == t[i - p[i] -1]; {
			p[i]++
		}
		if (i + p[i]) > right {
			idx = i
			right = i + p[i]
		}
		if p[i] > p[mIdx] {
			mIdx = i
		}
	}

	start := (mIdx - p[mIdx] - 1) / 2
	return s[start:start + p[mIdx]]
}

func preProcess(s string) string {
	t := make([]byte, len(s) * 2 + 3)
	t[0] = '^'
	t[len(t) - 1] = '$'

	for i, j := 1, 0; i < len(t) - 1; i++ {
		t[i] = '#'
		if (i % 2 == 0) {
			t[i] = s[j]
			j++
		}
	}
	return string(t)
}

func min(n1, n2 int) int {
	if n1 > n2 {
		return n2
	} else {
	  	return n1
	}
}